src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/update/actions/InstallNonIngressRulesAction.java

```diff
@@ -84,16 +87,40 @@ protected void perform(State from, State to,
     }
 
     private Collection<FlowSegmentRequestFactory> buildCommands(FlowCommandBuilder commandBuilder,
-                                                                CommandContext context, Flow flow,
-                                                                FlowPath path, FlowPath oppositePath,
-                                                                FlowUpdateFsm.EndpointUpdate endpointUpdate) {
-        switch (endpointUpdate) {
+                                                                FlowUpdateFsm stateMachine, Flow flow,
+                                                                FlowPath path, FlowPath oppositePath) {
+        CommandContext context = stateMachine.getCommandContext();
+        switch (stateMachine.getEndpointUpdate()) {
             case BOTH:
                 return new ArrayList<>(commandBuilder.buildEgressOnly(context, flow, path, oppositePath));
             case SOURCE:
-                return new ArrayList<>(commandBuilder.buildEgressOnlyOneDirection(context, flow, oppositePath, path));
+                switch (stateMachine.getFlowLoopOperation()) {
```

Comment:
A switch statement inside another switch is very confusing. Pls, consider refactoring - e.g. move into a method.

Reply:
reworked